File: py/data_structures/Stack.py
class StackNode:
    def __init__(self, val):
        self.val = val
        self.next = None 
        self.prev = None 

class Stack:
    def __init__(self):
        self.tail = None

    def push(self,val):
        new_node = StackNode(val)
        if self.tail:
            self.tail.next = new_node
            new_node.prev = self.tail 
            self.tail = new_node
        else:
            self.tail = new_node

    def pop(self):
        if self.tail:
            new_tail = self.tail.prev
            old_tail_val = self.tail.val 
            new_tail.next = None
            self.tail = new_tail 
            return old_tail_val
    
    def top(self):
        if self.tail:
            return self.tail.val

    def __len__(self):
        length = 0
        node = self.tail 
        while node:
            length += 1
            node = node.next
        return length 

File: js/data structures/Stack.js
import {DoubleLinkedListNode} from '../data structures/utils/LinkedList';

/**
 * This class represents the data structure known as a Stack, implemented using doubly linked list
 * nodes. 
 */
class Stack {
    constructor() {
        this.tail = null; 
    }

    push(val) {
        let node = new DoubleLinkedListNode(val);
        if (this.tail) {
            this.tail.next = node;
            node.prev = this.tail;
            this.tail = node; 
        }
        else {
            this.tail = node; 
        }
    }

    pop() {
        if (this.tail) {
            let oldTailVal = this.tail.val;
            if (this.tail.prev) {
                let newTail = this.tail.prev; 
                newTail.next = null;
                this.tail = newTail;
            }
            else {
                this.tail = null;
            }
            return oldTailVal; 
        }
        else {
            throw Error("pop from empty stack");
        }
    }

    top() {
        if (this.tail) {
            return this.tail.val;
        }
        else {
            throw Error("empty stack")
        }
    }

    length() {
        let size = 0;
        let node = this.tail; 
        while (node) {
            size++;
            node = node.prev;
        }
        return size; 
    }

}

export {Stack}; 